docs/testing-guide/contract-testing.md

```diff
@@ -0,0 +1,190 @@
+
+```{testcode}
+import algopy
+import pytest
+from algopy_testing import AlgopyTestContext, algopy_testing_context
+
+class CounterContract(algopy.Contract):
+    def __init__(self):
+        self.counter = algopy.UInt64(0)
+
+    @algopy.subroutine
+    def increment(self):
+        self.counter.set(self.counter.value + algopy.UInt64(1))
+        return algopy.UInt64(1)
+
+    @algopy.arc4.baremethod
+    def approval_program(self):
+        return self.increment()
+
+    @algopy.arc4.baremethod
+    def clear_state_program(self):
+        return algopy.UInt64(1)
+
+@pytest.fixture()
+def context():
+    with algopy_testing_context() as ctx:
+        yield ctx
+        ctx.reset()
+
+def test_counter_contract(context: AlgopyTestContext):
+    # Instantiate contract as you normally would with a regular python class
+    contract = CounterContract()
+
+    # Access the default creator account from the test context (created by default on test context creation)
+    user = context.default_creator
+
+    # Set the transaction group to emulate the application call
+    context.set_transaction_group(
+        gtxn=[
+            context.any_application_call_transaction(
+                sender=user,
+                app_id=context.any_application(),
+                app_args=[algopy.Bytes(b"increment")],
+            ),
+        ],
+        active_transaction_index=0,
+    )
+
+    # Invoke approval program
+    result = contract.approval_program()
+
+    # Assert approval program result
+    assert result == algopy.UInt64(1)
+
+    # Assert counter value
+    assert contract.counter.value == algopy.UInt64(1)
+
+    # Test clear state program
+    assert contract.clear_state_program() == algopy.UInt64(1)
+```
+
+## Defer contract method invocation
+
+You can create deferred application calls for more complex testing scenarios where order of transactions needs to be controlled:
+
+```python
+def test_deferred_call(context):
+    contract = MyARC4Contract()
+
+    extra_payment = context.any.txn.payment()
+    extra_asset_transfer = context.any.txn.asset_transfer()
+    implicit_payment = context.any.txn.payment()
+    deferred_call = context.txn.defer_app_call(contract.some_method, implicit_payment)
+
+    with context.txn.create_group([extra_payment, deferred_call, extra_asset_transfer]):
+        result = deferred_call.submit()
+
+    print(context.txn.last_group) # [extra_payment, implicit_payment, app call, extra_asset_transfer]
+```
+
+A deferred application call, assembles the application call transaction and can be later executed by calling `.submit()` method on the deferred application call instance. As shown on the example, it can also be passed to the transaction group creation context manager to be executed as part of the transaction group. In such cases if there are more than an application call inside the deferred objects, they will be executed in the order they were added to the transaction group.
```

Comment:
This isn't quite correct, only the method passed when calling `defer_app_call` will be executed on `submit()`